### PR TITLE
[All bundles] added postgres support

### PR DESCRIPTION
**Pull request by IntegraConsult**
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes
| BC breaks?    | No
| Deprecations? | No
| Fixed tickets | 

Upgraded Kunstmaan 5.7.1 so it supports posgresql. A lot of effort was put into keeping the code compatible with the curent supportted databases, so I do not expect it to break current functionality.
Following test were done before this pull request:
- manual click through the default application => no errors found
- kunstmaan php unit test suite => no errors found
-

### CODE REVIEW

**Review by ProfessorKuma — CHANGES_REQUESTED** [review]
Hi @, your PR needs some changes
* Your answer if this PR is a bugfix seems to be incorrect.
* Your answer if this PR is a new feature seems to be incorrect.
* Your answer if this PR is BC breaking seems to be incorrect.
* Your answer if this PR is has deprecations seems to be incorrect.
* your PR title should look like `[SomeBundle] Fixed some code`

**Review by ProfessorKuma — DISMISSED** [review]
Hi @, your PR needs some changes
* This PR seems to need a milestone of a minor release.

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php
Comment (acrobat) [inline src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php:229]: Why is the whole query duplicated? Can't the solution from [#2811](https://github.com/Kunstmaan/KunstmaanBundlesCMS/pull/2811/files#diff-2ae42119f573a6a61e26009641aa186d97e0f77b951dc02385d67d9daa075854) be used?

File: src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php
Comment (acrobat) [inline src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php:75]: Is this check necessary? I saw that in [#2811](https://github.com/Kunstmaan/KunstmaanBundlesCMS/pull/2811/files#diff-ad01d84409afe7ac963af2ce17fd3a076a100eed9c35b6d5e13c69cc75ef46df) the order by was removed and that seems like the right solution as a count with order by isn't needed I think

File: src/Kunstmaan/MediaBundle/Repository/FolderRepository.php
Comment (acrobat) [inline src/Kunstmaan/MediaBundle/Repository/FolderRepository.php:108]: Why is this check necessary?  The mysql variant can also use the new parameter, so we can avoid duplicated queries.

File: src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php
Comment (acrobat) [inline src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php:177]:
```suggestion
            ->andWhere('nt.online = :onlineTrue')
            ->setParameter('onlineTrue', true)
        ;
```

File: src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php
Comment (acrobat) [inline src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php:332]:
```suggestion
```

File: src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php
Comment (acrobat) [inline src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php:537]: Keep the indentation as it was before

File: src/Kunstmaan/TranslatorBundle/Repository/TranslationRepository.php
Comment (acrobat) [inline src/Kunstmaan/TranslatorBundle/Repository/TranslationRepository.php:65]: Can you remove this comment. It makes sense in this PR and the context of adding postgres support, but this info is not needed after it's merged

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php
Comment (acrobat) [inline src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php:63]: When the query is not duplicated any more I would inline this check in the query method, there is no need to have this as a class property

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php
Comment (acrobat) [inline src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php:69]: See comments in `AclHelper.php` about duplicate query

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclWalker.php
Comment (acrobat) [inline src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclWalker.php:39]: Cleanup this var_dump

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php
Comment (acrobat) [inline src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php:252]: Cleanup this var_dump

**Review by acrobat — CHANGES_REQUESTED** [review]
Hi @IntegraConsult,

Thanks for the PR, the effort is appreciated!
I did a first review and left some comments. As a general comment I would like to keep the query parameters "neutral". So `:onlineTrue` -> `:online`, this to avoid confusion if the value would change in the future but the parameter not.

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php:229]: As a matter of fact I initially did use the solution from #2811 and tested it with one of KBC bank core products. 2811 breaks access control in mysql, due to various issues (amongst which the databaseprefix I think). In order to keep things backwards compatibility with mysql I thiought it prudent to implement a dedicated method for postgres. However as you can see my implementation keeps many of the good stuff of 2811. 
The alternative would have been to place some more (if (dtatabaseplatform==) {}else{} statements in the opriginal code. I do not think that improves code readibility, but perhaps thsi si a matter of tasete

File: src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php:75]: Yes, I believe this check to be neccesary. In #2811 orderby was removed becaused it creates postgres errors when used in conjunction with COUNT as it is here. Simply removing the order by - because its not neccesary- from the original kunstmaan code is unfortunately not enogh, this may work for simple queries that are provided by querybuilder. However when queries provided by queryBuilder get more complex and contain orderBys in themselves we are back to square one. Hence the check

File: src/Kunstmaan/MediaBundle/Repository/FolderRepository.php
Comment (IntegraConsult) [reply src/Kunstmaan/MediaBundle/Repository/FolderRepository.php:108]: Good point. In hindsight I do not know why I did that either. Perheaps being a little overzealous trying to protect backward compatibility. No matter. check will be removed.

File: src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php
Comment (IntegraConsult) [reply src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php]: OK, done

**Comment by IntegraConsult** [comment]
> Hi @IntegraConsult,
> 
> Thanks for the PR, the effort is appreciated!
> I did a first review and left some comments. As a general comment I would like to keep the query parameters "neutral". So `:onlineTrue` -> `:online`, this to avoid confusion if the value would change in the future but the parameter not.

See your point, will update

File: src/Kunstmaan/TranslatorBundle/Repository/TranslationRepository.php
Comment (IntegraConsult) [reply src/Kunstmaan/TranslatorBundle/Repository/TranslationRepository.php]: done

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php]: Good point ! done

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclWalker.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclWalker.php:39]: done

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php]: done

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php]: Agree, done

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclWalker.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclWalker.php]: done

File: src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php:75]: Sometimes the queryBuilder query itself may contain orderBy clauses which will reek havoc in postgress when used in conjuction with COUNT DISTINCT. Therefore it is not sufficient to simply remove ones Kunstmaans own orderBy clause, but also the orderby clauses of the querybuilder query itself.  IMO the reason that the orderBy was removed from the kunstmaan code was not that it is not neccesary, but in postgres it is detrimental.

File: src/Kunstmaan/MediaBundle/Repository/FolderRepository.php
Comment (IntegraConsult) [reply src/Kunstmaan/MediaBundle/Repository/FolderRepository.php]: Absolutuetly right: ccheck is taken out

File: src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php
Comment (IntegraConsult) [reply src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php]: done

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php]: done

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclNativeHelper.php]: Same answer as above applies. The original 2811 was tested against one of KBC core system and caused crashes all over the place. To make abolutely sure backward compatibility with msql is guearanted I opted to do a separate method for postgres. This is not critical/essential however and could be refactored to a single method with some mor if(){}else{} clauses. Apart from processing a few more ifs, IMO this makes the code less readable, so I would not do it ifg the product was my own. However I also see that this can be a matter of taste.

**Comment by IntegraConsult** [comment]
Hello Jeroen, Ad Langenkamp here

I went through review comments, and made the appropriate code changes. Some
of the original 2811 PR needed to be reworked so they would work in a more
general case. This particularly pertains to AclHelper, AclNativeHelper and
DoctrineDBALAdapter.

Can you please have a look and see if the changes are OK as they are, or
perhaps a little more needs to be done?.

Thank you

Ad Langenkamp

(currently working for ICAPPS/KBC bank belgium)


Op wo 17 feb. 2021 om 20:32 schreef Jeroen Thora <notifications@github.com>:

File: src/Kunstmaan/AdminBundle/Repository/ExceptionRepository.php
Comment (acrobat) [inline src/Kunstmaan/AdminBundle/Repository/ExceptionRepository.php:32]: I don't know if it's needed to pass the 3th parameter here (I know it brings a speed improvement for `\DateTime` objects), so we should decide to add or not add it. Because now it's a mix of both, so if it's not required I would suggest not adding it.

EDIT: or is it here required because it's a parameter for an update field?

File: src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php
Comment (acrobat) [reply src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php:75]: Yes I know postgres has issues with that, but what I meant was can't this be done for "all" (mysql and postgress) platforms? So we remove the general orderBy and reset the OrderBy on the query object.

File: src/Kunstmaan/MediaBundle/Repository/FolderRepository.php
Comment (acrobat) [inline src/Kunstmaan/MediaBundle/Repository/FolderRepository.php:111]: Same question as in `src/Kunstmaan/AdminBundle/Repository/ExceptionRepository.php`. If it's not required I would remove the 3th parameter as we don't set it anywhere else.

File: src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php
Comment (acrobat) [inline src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php:282]: See comment/question above about the 3th parameter

File: src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php
Comment (acrobat) [reply src/Kunstmaan/AdminBundle/Helper/Security/Acl/AclHelper.php:223]: I will look into this myself to see what the posibilities are. But my gut feeling is that this is not ideal as this just duplicates the full code. Maybe just inner parts need to be duplicated. But let's see what an ideal solution can be here!

**Review by acrobat — CHANGES_REQUESTED** [review]
Thanks for the updated parameter names, they look good to me! I left some extra comments/questions before actually testing this PR in some real life projects.

File: src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminListBundle/Helper/DoctrineDBALAdapter.php]: OK, made the change also for non postgress and extensively retested that against mysql database. You are right. This works also for mysql, so no if statement needed and cleaner code. Thanks!

File: src/Kunstmaan/AdminBundle/Repository/ExceptionRepository.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Repository/ExceptionRepository.php]: Your right, this is not consistent, have removed them

File: src/Kunstmaan/MediaBundle/Repository/FolderRepository.php
Comment (IntegraConsult) [reply src/Kunstmaan/MediaBundle/Repository/FolderRepository.php]: same answer: PARAM_BOOL's have been removed

Comment (IntegraConsult) [reply src/Kunstmaan/MediaBundle/Repository/FolderRepository.php]: Yep, has been removed

File: src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php
Comment (IntegraConsult) [reply src/Kunstmaan/NodeBundle/Repository/NodeTranslationRepository.php]: parameter has beeen removed

File: src/Kunstmaan/AdminBundle/Repository/ExceptionRepository.php
Comment (IntegraConsult) [reply src/Kunstmaan/AdminBundle/Repository/ExceptionRepository.php]: parameter was removed

**Comment by IntegraConsult** [comment]
As you can see I made the changes you requested. Thanks for posing your review remarks in such a constructive manner. It really helped me!
Ad

**Review by acrobat — CHANGES_REQUESTED** [review]
@IntegraConsult I've created a PR against your fork with extra fixes. Can you review and verify in your project? 

See: https://github.com/IntegraConsult/KunstmaanBundlesCMS/pull/1

**Review by acrobat — CHANGES_REQUESTED** [review]

File: src/Kunstmaan/NodeBundle/Repository/NodeRepository.php
Comment (acrobat) [inline src/Kunstmaan/NodeBundle/Repository/NodeRepository.php:358]: This is something I would like to solve in the query, as it could potentially a huge array to be looped over on large websites. If we can "pre-filter" these results in the query, this heavy lifting is done by the database

**Review by acrobat — CHANGES_REQUESTED** [review]

File: src/Kunstmaan/NodeBundle/Repository/NodeRepository.php
Comment (acrobat) [inline src/Kunstmaan/NodeBundle/Repository/NodeRepository.php:280]:
```suggestion
                    return sprintf('COALESCE(%s, %s)', $falseValue, $trueValue);
```

As discussed on slack!

**Review by acrobat — APPROVED** [review]

**Comment by acrobat** [comment]
Thanks @IntegraConsult!

**Pull request merged**
